### PR TITLE
Code to convert encoder ticks to odometry

### PR DESCRIPTION
**Pull request by rachelzh9**
publishes to odom_data_euler

### CODE REVIEW

File: src/odom_node.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  quatOdom.pose.pose.orientation.z = q.z();
+  quatOdom.pose.pose.orientation.w = q.w();
+  quatOdom.twist.twist.linear.x = odomNew.twist.twist.linear.x;
+  quatOdom.twist.twist.linear.y = odomNew.twist.twist.linear.y;
+  quatOdom.twist.twist.linear.z = odomNew.twist.twist.linear.z;
+  quatOdom.twist.twist.angular.x = odomNew.twist.twist.angular.x;
+  quatOdom.twist.twist.angular.y = odomNew.twist.twist.angular.y;
+  quatOdom.twist.twist.angular.z = odomNew.twist.twist.angular.z;
+ 
+  for(int i = 0; i<36; i++) {
```
Comment (jsongCMU) [inline src/odom_node.cpp]: Comment to explain this?

File: src/odom_node.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  }
+ 
+  // Make sure theta stays in the correct range
+  if (odomNew.pose.pose.orientation.z > PI) {
+    odomNew.pose.pose.orientation.z -= 2 * PI;
+  }
+  else if (odomNew.pose.pose.orientation.z < -PI) {
+    odomNew.pose.pose.orientation.z += 2 * PI;
+  }
+  else{}
```
Comment (jsongCMU) [inline src/odom_node.cpp]: What's this for?

File: src/odom_node.cpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+}
+ 
+// Update odometry information
+void update_odom(ros::Publisher &odom_data_pub) {
+ 
+    long delta_pos_left, delta_pos_right;
+    float delta_left, delta_right, delta_theta, theta2;
+    float delta_x, delta_y;
+
+    delta_pos_left = pos_left - pos_left_prev;
```
Comment (jsongCMU) [inline src/odom_node.cpp]: Any references or documents of where these equations come from?

**Comment by jsongCMU** [comment]
Any testing to confirm this works? Maybe move robot using teleop and see how much distance it travels, and compare with calculated delta x, delta y, delta theta?

File: src/odom_node.cpp (hunk 2 above)
Comment (rachelzh9) [reply src/odom_node.cpp]: keep orientation angles between [-pi, pi]

File: src/odom_node.cpp (hunk 3 above)
Comment (rachelzh9) [reply src/odom_node.cpp]: khepera IV toolbox xd

File: src/odom_node.cpp (hunk 1 above)
Comment (rachelzh9) [reply src/odom_node.cpp]: Covariance matrix is 6x6, loop through to set diagonal values

**Comment by indraneelpatil** [comment]
LGTM! Great job @rachelzh9 !!

**Pull request merged**